File: front/src/pages/user/validators.ts
import { z } from 'zod'

export const createUserValidator = z.object({
  username: z.string().min(1, 'Username is required'),
  firstname: z.string().optional(),
  lastname: z.string().optional(),
  email: z.union([z.string().email(), z.literal('')]).optional(),
  email_verified: z.boolean().optional(),
})

export const updateUserValidator = z.object({
  username: z.string().min(1, 'Username is required'),
  firstname: z.string().optional(),
  lastname: z.string().optional(),
  enabled: z.boolean().optional(),
  email: z.union([z.string().email(), z.literal('')]).optional(),
  email_verified: z.boolean().optional(),
  required_actions: z.array(z.string()).optional(),
})


export type CreateUserSchema = z.infer<typeof createUserValidator>
export type UpdateUserSchema = z.infer<typeof updateUserValidator>

File: api/src/application/http/user/validators.rs
use serde::{Deserialize, Serialize};
use utoipa::ToSchema;
use uuid::Uuid;
use validator::Validate;

#[derive(Debug, Serialize, Deserialize, Validate, ToSchema)]
pub struct ResetPasswordValidator {
    #[serde(default)]
    pub temporary: bool,

    #[serde(default)]
    pub credential_type: String,

    #[validate(length(min = 1, message = "value is required"))]
    #[serde(default)]
    pub value: String,
}

#[derive(Debug, Serialize, Deserialize, Validate, ToSchema)]
pub struct CreateUserValidator {
    #[validate(length(min = 1, message = "username is required"))]
    #[serde(default)]
    pub username: String,

    #[serde(default)]
    pub firstname: Option<String>,

    #[serde(default)]
    pub lastname: Option<String>,

    #[serde(default)]
    pub email: Option<String>,

    #[serde(default)]
    pub email_verified: Option<bool>,
}

#[derive(Debug, Serialize, Deserialize, Validate, ToSchema)]
pub struct BulkDeleteUserValidator {
    #[serde(default)]
    pub ids: Vec<Uuid>,
}

#[derive(Debug, Serialize, Deserialize, Validate, ToSchema)]
pub struct UpdateUserValidator {
    #[serde(default)]
    pub firstname: Option<String>,

    #[serde(default)]
    pub lastname: Option<String>,

    #[serde(default)]
    pub email: Option<String>,

    #[serde(default)]
    pub email_verified: Option<bool>,

    #[serde(default)]
    pub enabled: Option<bool>,

    #[serde(default)]
    pub required_actions: Option<Vec<String>>,
}
